Sources/Basic/Path.swift

```diff
@@ -0,0 +1,473 @@
+
+
+/// Private function that returns a string containing the home directory of the
+/// effective user.  This should definitely move out into a public place, but
+/// we'll need to debate the names, etc, etc.
+/// FIXME:  Should this function return an optional or throw an error if there
+/// is a problem with the current user id?
+private func gethomedir() -> String {
+    /// First respect the `HOME` environment variable, if it's set.
+    if let path = POSIX.getenv("HOME") {
+        return path
+    }
+    // Otherwise check effective user id, falling back on actual user id.
+    var uid = geteuid()
+    if uid == 0 { uid = getuid() }
+    // Look up the user in the account database.
+    let pwd = getpwuid(uid)
+    if let cdir = pwd?.pointee.pw_dir {
+        // We found the user, so construct a string from the pw_dir.
+        if let dir = String(validatingUTF8: cdir) {
+            return dir
+        }
+        // FIXME: How should we recover from bad UTF-8 in the home directory?
+        //        Should we just throw an error and punt the problem upwards?
+    }
+    // FIXME: What should we return if all else fails?
+    return "/tmp/~"
+}
+
+
+/// Private function that returns a string containing the home directory of the
+/// name user.  This should definitely move out into a public place, but we'll
+/// need to debate the names, etc, etc.
+/// FIXME:  Should this function return an optional or throw an error if there
+/// is no user with the given name?
+private func gethomedir(user: String) -> String {
+    // Look up the user in the account database.
+    let pwd = getpwnam(user)
+    if let cdir = pwd?.pointee.pw_dir {
+        // We found the user, so construct a string from the pw_dir.
+        if let dir = String(validatingUTF8: cdir) {
+            return dir
+        }
+        // FIXME: How should we recover from bad UTF-8 in the home directory?
+        //        Should we just throw an error and punt the problem upwards?
+    }
+    // FIXME: What should we return if all else fails?
+    return "/tmp/~\(user)"
+}
+
+
+/// Convenience properties that access the file system, making it convenient to
+/// get information about a file system entity at a particular path.  Note that
+/// none of these functions is allowed to mutate the file system.
+extension AbsolutePath {
+    // FIXME: Factor out a method to return the stat record, and then move the
+    //        other methods to the stat record.  That makes it easier to avoid
+    //        blithely stat:ing over and over again if multple properties are
+    //        checked (e.g. `isFile || isSymlink` etc).
+    
+    public var isDirectory: Bool {
+        var mystat = stat()
+        let rv = lstat(self.asString, &mystat)
+        return rv == 0 && (mystat.st_mode & S_IFMT) == S_IFDIR
+    }
+    
+    public var isFile: Bool {
+        var mystat = stat()
+        let rv = lstat(self.asString, &mystat)
+        return rv == 0 && (mystat.st_mode & S_IFMT) == S_IFREG
```

Comment:
Yes.  Or rather, this whole API needs to be redone, as noted in the FIXME.  The user needs to be able to choose whether or not to examine the entity at the path vs any resolved symlink.  The old code which always used stat() was definitely wrong, since it could never return true for isSymlink.